python/cudf_polars/cudf_polars/experimental/scheduler.py

```diff
@@ -0,0 +1,171 @@
+
+    Parameters
+    ----------
+    key
+        Root key we want to extract.
+    graph
+        The full task graph.
+
+    Returns
+    -------
+    List of other keys needed to extract ``key``.
+    """
+    maybe_task = graph[key]
+    return [
+        k
+        for k in (
+            maybe_task[1:]
+            if istask(maybe_task)
+            else [maybe_task]  # maybe_task might be a key
+        )
+        if is_hashable(k) and k in graph
+    ]
+
+
+def toposort(graph: Graph, dependencies: Mapping[Key, list[Key]]) -> list[Key]:
+    """Return a list of task keys sorted in topological order."""
+    # Stack-based depth-first search traversal. This is based on Tarjan's
+    # method for topological sorting (see wikipedia for pseudocode)
+    ordered: list[Key] = []
+    completed: set[Key] = set()
+    seen: set[Key] = set()
+
+    for key in graph:
+        if key in completed:
+            continue
+        nodes: list[Key] = [key]
+        while nodes:
+            # Keep current node on the stack until all descendants are visited
+            current = nodes[-1]
+            if current in completed:  # pragma: no cover
+                # Already fully traversed descendants of current
+                nodes.pop()
+                continue
+            seen.add(current)
+
+            # Add direct descendants of current to nodes stack
+            next_nodes = []
+            for nxt in dependencies[current]:
+                if nxt not in completed:
+                    if nxt in seen:  # pragma: no cover
+                        # Cycle detected!
+                        raise RuntimeError(
+                            f"Cycle detected in the task graph! Key: {nxt}"
+                        )
+                    next_nodes.append(nxt)
+
+            if next_nodes:
+                nodes.extend(next_nodes)
+            else:
+                # Current has no more descendants to explore
+                ordered.append(current)
+                completed.add(current)
+                seen.remove(current)
+                nodes.pop()
+
+    return ordered
+
+
+def synchronous_scheduler(
+    graph: Graph,
+    key: Key,
+    *,
+    cache: MutableMapping | None = None,
+) -> Any:
+    """
+    Execute the task graph for a given key.
+
+    Parameters
+    ----------
+    graph
+        The task graph to execute.
+    key
+        The final output key to extract from the graph.
+    cache
+        Intermediate-data cache.
+
+    Returns
+    -------
+    Executed task-graph result for ``key``.
+    """
+    if key not in graph:  # pragma: no cover
+        raise KeyError(f"{key} is not a key in the graph")
+    if cache is None:
+        cache = {}
+
+    refcount: defaultdict[Key, int] = defaultdict(int)
+    dependencies: dict[Key, list] = {}
+    for k in graph:
+        dependencies[k] = required_keys(k, graph)
+        for dep in dependencies[k]:
+            refcount[dep] += 1
+
+    for k in toposort(graph, dependencies):
```

Comment:
Just to note, I don't think we're making graphs like this right now, but this potentially does much more work than necessary in execution. What we want to do is extract a particular key, which means we need to execute all of the dependencies _of that key_ in some topological order. Here we're executing all of the _graph_ in topological order, which may be more than we need to.

I don't think we should fix this now though.

Reply:
Yes absolutely. The current algorithm leads to a dramatic performance penalty if we have anything "extra" in the task graph. We can absolutely revisit the algorithm in a follow-up.

Reply:
I assumed that in our particular use case we would only ever have a graph containing exactly what we need, but if at some point we want to make that more flexible I agree that this needs fixing eventually.